### ISSUE TITLE
unigned-signatures: incorrect merge proposal of optional parameters

### ISSUE THREAD

**Issue by stephanedr** [palantir/tslint#2088]
### Bug Report

- __TSLint version__: `3.4.1`
- __TypeScript version__: `2.5.1`
- __Running TSLint via__: VSCode

#### TypeScript code being linted

```ts
function test(): void;
function test(a: number): void;
function test(a?: number): void {}
```

#### Actual behavior
'These overloads can be combined into one signature with an optional parameter. (unified-signatures)'
at: '2,15'
'These overloads can be combined into one signature with an optional parameter. (unified-signatures)'
at: '3,15'  --> due to #2016.

#### Expected behavior
The 2 signatures are not mergeable because `test(a?: number): void` allows `test(undefined)` whereas the given code doesn't with `--strictNullChecks`.

**Comment by andy-ms** [comment]
If it didn't warn in this situation the rule wouldn't do much at all. Do you really need to prevent people from calling `test(undefined)` exactly? The example with overloads will also prevent `test(Math.random() > 0.5 ? 1 : undefined);`, but that would presumably work at runtime, so why not allow it?

**Comment by stephanedr** [comment]
My case is a bit more complex (more parameters, so combinations) than the simplest test case I provided to show the issue.
And yes I want to prevent calling `test(undefined)`.
To me:
```
function test(a: number): void;
function test(a: string): void;
```
can indeed be merged to `function test(a: number | string): void`. So this rules applies here.

But to me this rule shouldn't consider that no parameter (`test()`) is equivalent to an optional parameter.
Maybe this can be controlled through an option?

**Comment by andy-ms** [comment]
Maybe -- the rule does have separate checking for `signaturesDifferBySingleParameter` and `signaturesDifferByOptionalOrRestParameter`.  I just don't understand why you need to define a function that could take 0 or 1 arguments, but can't take an optional argument. Seems to needlessly make people replace `f(arg)` with `arg ? f(arg) : f()`.

**Comment by stephanedr** [comment]
As I mentioned, my case is more complex, and I capture arguments through a rest argument.
Something that can be simplified to:
```
function test(): void;
function test(a: number): void;
function test(...a: any[]): void { }
```
The problem is that `arguments.length == 1` with `test(undefined)`. That I want to avoid.
But I agree this may be a too specific situation (that I can avoid by either implementing the function differently or by adding a disable rule comment).

Forget about this request.
